Add node.js client library

g-k opened the pull request:
This adds the node.js client library I started last week to integrate with the FxA customs server.

We're actually planning on [sending violations](https://github.com/mozilla-services/tigerblood/issues/16) instead of setting reputation directly, but the existing endpoints might still be useful for testing or an admin panel.

Commit:
Add node.js client library

jrgm commented on node-js-client/scripts/tap-coverage.js:
```diff
@@ -0,0 +1,38 @@
+
+/* eslint-disable no-console */
+if (!process.env.NO_COVERAGE) {
+  var ass = require('ass').enable( {
```
+1 for coverage, and I realize this is mimicing style in customs-server, but the `ass` module is pretty deprecated at the moment. Maybe have a look at how some other fxa modules were ported over to use `nyc` (which is `istanbul` based and can plug into the `coveralls` CI service).

g-k replied:
Thanks @jrgm looks like fxa-auth-mailer is using nyc, so I'll check that out and get this updated.

g-k replied:
switched to nyc and coveralls nice to have branch coverage.

Can you add the COVERALLS_REPO_TOKEN environment variable to the CircleCI config or should I have a separate travis build?

g-k commented:
r? @vbudhram

Should this library:
- take a timeout param
- use promises instead of callbacks
- shrinkwrap its dependencies

?

Commit:
switch testing to nyc and coveralls

per jrgm's comment: https://github.com/mozilla-services/tigerblood/pull/18#discussion_r84790091

Commit:
remove unused eslint fxa config

vbudhram commented:
@g-k Haven't had a chance to fully review this (should by tomorrow) but here are my thoughts on your other questions

> take a timeout param

I believe so, currently the fxa-js-client has a 30 second [timeout](https://github.com/mozilla/fxa-js-client/blob/f98d579f9ec834639fa2815adbbb6d8c7dcf0d95/client/lib/request.js#L24).

> use promises instead of callbacks

The custom-server uses promises so I think using promises might make for a littler easier integration. The section of the customs server that you will probably be most interested in is [here](https://github.com/mozilla/fxa-customs-server/blob/master/lib/server.js#L188).

> shrinkwrap its dependencies

Our current fxa-js-client does not shrinkwrap dependencies so I don't think that is necessary. However, I will follow up with the team to confirm that this shouldn't be shrinkwrapped.

vbudhram commented on node-js-client/.eslintrc:
```diff
@@ -0,0 +1,2 @@
+rules:
```
Up to you, but you can use [eslint-config-fxa](https://github.com/pdehaan/eslint-config-fxa) and extend the `fxa` config. This is a similar setup that the [fxa-js-client](https://github.com/mozilla/fxa-js-client/blob/master/.eslintrc#L2) has.

vbudhram commented on node-js-client/lib/client.js:
```diff
@@ -0,0 +1,140 @@
+  if (!Object.prototype.hasOwnProperty.call(config, 'host')) {
+    throw new Error('Missing required param host for IP Reputation Client.')
+  } else if (!Object.prototype.hasOwnProperty.call(config, 'port')) {
+    throw 'Missing required param port for IP Reputation Client.'
```
Did you mean to `throw new Error` here and below?

vbudhram replied:
Was this going to be used only server-side? If so, could possibly use [joi](https://github.com/hapijs/joi) to do object validation.

g-k replied:
Yeah, I did mean to throw an error. This will be server-side only, so I switched to joi. Much nicer.

vbudhram commented on node-js-client/lib/client.js:
```diff
@@ -0,0 +1,140 @@
+ * @param {Object} config
+ *   @param {String} config.host
+ *   @param {Number} config.port
+ *   @param {id} config.id id for the HAWK header
```
Do you mean `{String}` for id and key for header?

g-k replied:
Yep, fixed.

vbudhram commented on node-js-client/lib/client.js:
```diff
@@ -0,0 +1,140 @@
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    body: JSON.stringify({'ip': ip, 'reputation': reputation})
```
According to request docs, I believe you can also use `json` if you pass object and thus avoid having to do the `JSON.stringify`

> json - sets body to JSON representation of value and adds Content-type: application/json header. Additionally, parses the response body as JSON.

g-k replied:
sweet

https://github.com/mozilla-services/tigerblood/pull/18/commits/d50650c0273d75b92a484b545d2e6b3624d91cf6

vbudhram commented on node-js-client/test/local/reputation_service_client_tests.js:
```diff
@@ -0,0 +1,119 @@
+  function (t) {
+    client.update('127.0.0.1', 5, function (error, response, body) {
+      t.equal(error, null)
+      t.equal(response.statusCode, 200)
```
Is it correct that the service returns `200` if the item does not exist and can not be updated?

g-k replied:
Yeah, I was running against a version before the fix for https://github.com/mozilla-services/tigerblood/issues/15 landed.

vbudhram commented on node-js-client/test/local/reputation_service_client_tests.js:
```diff
@@ -0,0 +1,119 @@
+  function (t) {
+    client.add('127.0.0.1', 50, function (error, response, body) {
+      t.equal(error, null)
+      t.equal(response.statusCode, 500)
```
Is a `500` the correct response here? Should there checks to ensure that the correct error message was sent?

g-k replied:
That is weird. opened a PR https://github.com/mozilla-services/tigerblood/pull/23

vbudhram commented on node-js-client/test/local/reputation_service_client_tests.js:
```diff
@@ -0,0 +1,119 @@
+test(
+  'removes reputation for existing IP',
+  function (t) {
+    client.remove('127.0.0.1', function (error, response, body) {
```
You can probably extend this test to perform `get` request for the IP, and check that it fails.

vbudhram reviewed (changes requested):
@g-k Just a couple of things code wise. My next step is to test client script with local version of Tigerblood.

Commit:
fix typos

Commit:
use fxa config; fix new Error throwing

Commit:
add timeout param

Commit:
fix test per issue #15

https://github.com/mozilla-services/tigerblood/issues/15

Commit:
validate config with joi

Commit:
fix no new for side effects lint

Commit:
switch to promises

Commit:
verify IP delete

Commit:
remove unused ext hawk param

Commit:
use the hawk request option

Commit:
remove extra json headers and stringify calls

Commit:
readme: update node.js client examples

g-k commented:
@vbudhram Thanks for the detailed feedback I think I got to everything.

Commit:
update per #23

Commit:
add violations

Commit:
Add yarn.lock

g-k commented:
Moved to a separate repo https://github.com/mozilla-services/ip-reputation-js-client per @autrilla's suggestion. This should make installing it easier too.
